### ISSUE TITLE
DC-LLP-017: BERT/MiniLM semantic task classifier (Tier 2 router)

### ISSUE THREAD

**Issue by radoxtech** [radoxtech/diricode#412]
## DC-LLP-017 — BERT/MiniLM Classifier (Tier 2 Router)

**Epic**: #3 (Provider Router — Transport Layer)
**Package**: `@diricode/dirirouter` (`packages/dirirouter/`)
**Iteration**: v2.0
**ADR**: [ADR-049](docs/adr/adr-049-llm-picker.md), [ADR-055](docs/adr/adr-055-diri-router-unified-package.md)
**Previously**: LLM Picker epic #389 (closed) → Sub-Epic #393 (closed)

---

### Goal

Implement the Tier 2 semantic task classifier using a BERT-class or MiniLM model running locally via ONNX Runtime. This tier provides fast, embedding-based task classification that narrows candidate models when Tier 1 heuristics are insufficient.

### Why This Matters

Tier 1 (rule-based heuristics from #519 POC) handles obvious cases: short questions → fast model, code blocks → coding model. But ambiguous prompts like "explain the trade-offs of microservices" need semantic understanding to classify correctly. Tier 2 fills this gap with <100ms inference, avoiding the latency cost of Tier 3 (full LLM reasoning).

#### Model Selection
- Primary: `all-MiniLM-L6-v2` (22M params, ~30ms inference on CPU)
- Fallback: `distilbert-base-uncased` (66M params, ~80ms inference)
- Format: ONNX for cross-platform, no Python dependency
- Model files stored in `packages/dirirouter/models/` (git-lfs or downloaded on first run)

#### Task Classification Categories
```typescript
type TaskCategory =
  | 'coding'        // Code generation, debugging, refactoring
  | 'research'      // Information gathering, analysis
  | 'planning'      // Project planning, architecture decisions
  | 'creative'      // Writing, brainstorming, ideation
  | 'quick_answer'  // Simple factual questions
  | 'reasoning'     // Complex multi-step logic
  | 'unknown';      // Falls through to Tier 3
```

#### Classification Pipeline
1. Tokenize input prompt (first 512 tokens)
2. Run ONNX inference → embedding vector
3. Compare against pre-computed category centroids (cosine similarity)
4. If max similarity > threshold (0.7): return category + confidence
5. If below threshold: return `unknown` → Tier 3 handles it

#### Integration with Picker Cascade
- Implements `PickerTier` interface: `classify(prompt: string) → { category: TaskCategory, confidence: number, latency: number }`
- Cascade order: Tier 1 (heuristics) → **Tier 2 (this)** → Tier 3 (Tiny LLM)
- Tier 2 only invoked when Tier 1 returns low confidence (<0.6)

### Acceptance Criteria

- [ ] ONNX Runtime integration loads MiniLM model successfully
- [ ] Classification latency < 100ms on standard CPU (measured via benchmark)
- [ ] Achieves >85% accuracy on DiriCode task type test set (coding, research, planning, creative, quick_answer, reasoning)
- [ ] Graceful fallback to Tier 1 heuristics on ONNX load failure or timeout
- [ ] `unknown` category correctly triggers Tier 3 escalation
- [ ] Category centroids stored in JSON config, updatable without code changes
- [ ] Memory footprint < 100MB for loaded model
- [ ] Unit tests cover: each category classification, confidence thresholds, fallback behavior, timeout handling

### Architecture Notes

- Lives in `packages/dirirouter/src/picker/tiers/tier2-classifier.ts`
- ONNX Runtime: `onnxruntime-node` package (CPU-only, no GPU requirement)
- Category centroids: `packages/dirirouter/config/task-centroids.json`
- Model files: `packages/dirirouter/models/minilm-l6-v2.onnx` (git-lfs tracked)
- Implements `PickerTier` interface from `packages/dirirouter/src/picker/types.ts`

### References

- [ADR-049](docs/adr/adr-049-llm-picker.md) — LLM Picker cascade architecture
- [ADR-055](docs/adr/adr-055-diri-router-unified-package.md) — Unified diri-router package
- #413 — Tier 3 (Tiny LLM router, next tier in cascade)
- #408 — Multi-dimensional model_scores (classification feeds into score-based selection)
- #422 — Auto re-training pipeline (re-trains this classifier)

### Dependencies

- Depends on: #408 (multi-dimensional scores for category→model mapping), ONNX Runtime npm package
- Required by: #422 (re-training pipeline), #435 (integration with routing stack)

**Comment by radoxtech** [comment]
## Update — Moved to Phase 2, blocked on training data

### What changed:
After analysis of the routing strategy and web research (RouteLLM, LiteLLM, vLLM Semantic Router), this issue is being **deferred to Phase 2**. The reasons:

1. **No training data exists yet** — BERT classifiers need labeled `(prompt, task_category, reward)` tuples. We have zero.
2. **`task.description` is optional** in current `DecisionRequest` — BERT can't classify empty input.
3. **Enhanced heuristics (DC-DR-039) + Embedding Router (DC-DR-040)** cover the immediate gap with zero training data requirement.
4. **DC-DR-041 (Routing Telemetry Collector)** will collect the decision data needed to train this classifier.

### New dependency chain:
```
DC-DR-039 (Enhanced Heuristic) → DC-DR-041 (Telemetry Collector) → DC-DR-040 (Embedding Router) → #412 (THIS — BERT, Phase 2)
```

### Prerequisites before this can start:
- [ ] DC-DR-041 deployed and collecting telemetry for ≥2 weeks
- [ ] Minimum 500 labeled decision pairs collected
- [ ] Playground schema (#458) landed — ensures `task.description` is required
- [ ] Benchmark baseline established from DC-DR-039 heuristic accuracy

### Changes to original spec:
- Training data source: telemetry from DC-DR-041 (not hypothetical `score_history`)
- Model: Consider Matrix Factorization (RouteLLM finding) as alternative to MiniLM
- Category centroids: bootstrapped from Embedding Router (DC-DR-040) experience

**Status: `status:blocked`** — waiting on telemetry data collection.

### Related:
- #413 (TinyLLM) — **CLOSED** as anti-pattern
- #641 — **CLOSED**, superseded by DC-DR-039/040/041
